middleware.go

```diff
@@ -0,0 +1,100 @@
+package jwt
+
+import (
+	"github.com/Digital-Insight-Technologies-Ltd/gin-jwt-cognito/models"
+	"net/http"
+
+	"github.com/gin-gonic/gin"
+	"github.com/go-errors/errors"
+	"github.com/golang-jwt/jwt/v5"
+	"github.com/google/uuid"
+)
+
+// UserClaimsKey is the key for user claims in our context
+const UserClaimsKey = "userClaims"
+
+// CognitoClaimsMiddleware is a middleware for Gin that extracts user claims from
+// AWS Cognito.
+func CognitoClaimsMiddleware() gin.HandlerFunc {
+	return func(c *gin.Context) {
+		tok, ok := c.GetQuery("Authorization")
+
+		if !ok {
+			c.AbortWithStatusJSON(http.StatusUnauthorized, gin.H{"error": "Missing Authorization header"})
+			return
+		}
+
+		claims, err := parseClaims(tok[7:])
+
+		if err != nil {
+			c.AbortWithStatusJSON(http.StatusUnauthorized, gin.H{"error": "Invalid user claims (OIDC: stage 1)"})
+			return
+		}
+
+		userCtx, err := constructUserContext(claims)
+
+		if err != nil {
+			c.AbortWithStatusJSON(http.StatusUnauthorized, gin.H{"error": "Invalid user claims (OIDC: stage 2)"})
+			return
+		}
+
+		// Add the claims to Gin's context
+		c.Set(UserClaimsKey, userCtx)
+
+		// Proceed with the middleware chain
+		c.Next()
+	}
+}
+
+func parseClaims(claimsHeader string) (jwt.MapClaims, error) {
+	tok, _, err := new(jwt.Parser).ParseUnverified(claimsHeader, jwt.MapClaims{})
```

Comment:
Optimize JWT parsing by reusing the `jwt.Parser` instance instead of creating a new one every time. Define a global parser if thread safety is ensured. [medium]